### PR TITLE
8284811

### PR DESCRIPTION
**Pull request by albertnetymk**
This refactoring around gc-locker consists of two commits:

1. Remove `_debug_jni_lock_count` because `verify_critical_count` already verifies #threads in critical regions. `_debug_jni_lock_count` doesn't introduce anything new.

2. Shrink the critical section associated with `JNICritical_lock` to accesses to `_needs_gc` and nothing else.

Test: tier1-6

### CODE REVIEW

**Comment by dholmes-ora** [comment]
I think I agree that _debug_jni_lock_count is not really adding anything ... but perhaps there is a subtlety we are overlooking. Digging into the history of the code may shed more light.

The rest of the changes are very hard to follow and seem to completely change the logic.

File: src/hotspot/share/gc/shared/gcLocker.cpp [hunk 1]
```diff
@@ -123,33 +106,30 @@ bool GCLocker::should_discard(GCCause::Cause cause, uint total_collections) {
          (_total_collections != total_collections);
 }
 
-void GCLocker::jni_lock(JavaThread* thread) {
+void GCLocker::stall_until_no_needs_gc(JavaThread* thread) {
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/gcLocker.cpp:109]: The increment of _jni_lock_count seems to have gotten lost.  I don't see how
this change can work without that.

File: src/hotspot/share/gc/shared/gcLocker.cpp [hunk 2]
```diff
@@ -123,33 +106,30 @@ bool GCLocker::should_discard(GCCause::Cause cause, uint total_collections) {
...
-  assert(thread->in_last_critical(), "should be exiting critical region");
-  MutexLocker mu(JNICritical_lock);
-  _jni_lock_count--;
-  decrement_debug_jni_lock_count();
-  thread->exit_critical();
-  if (needs_gc() && !is_active_internal()) {
+void GCLocker::ack_needs_gc_if_last_thread(JavaThread* thread) {
+  assert(!thread->in_critical(), "shouldn't currently be in a critical region");
+  assert(needs_gc(), "precondition");
+  jint old_value = Atomic::fetch_and_add(&_jni_lock_count, -1);
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/gcLocker.cpp:129]: Even assuming it's okay to modify _jni_lock_count outside the lock (which I'm not
at all sure about), I'm not convinced it's an improvement to do so.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
I don't understand how this change is supposed to work, and think it doesn't.
It would help to have some description of how the new code is (supposed to)
work, and how that's an improvement over the old code.

File: src/hotspot/share/gc/shared/gcLocker.inline.hpp [hunk 3]
```diff
@@ -30,29 +30,18 @@
...
-    if (needs_gc()) {
-      // jni_lock call calls enter_critical under the lock so that the
-      // global lock count and per thread count are in agreement.
-      jni_lock(thread);
-      return;
-    }
-    increment_debug_jni_lock_count();
+  if (!thread->in_critical() && needs_gc()) {
+    stall_until_no_needs_gc(thread);
+    assert(!needs_gc(), "postcondition");
```
[the remark this answers is not in the document]
Comment (albertnetymk) [reply src/hotspot/share/gc/shared/gcLocker.inline.hpp:35]: A thread running between the end of the stall and the end of the JNI API (e.g. `GetStringCritical`) will prevent safepoint from being reached and `_needs_gc` is set inside a safepoint, so this assertion is correct I think.

PS: The doc in `SafepointSynchronize::arm_safepoint` helped me greatly to understand when a safepoint is possible.

File: src/hotspot/share/gc/shared/gcLocker.cpp (hunk 1 above)
Comment (albertnetymk) [reply src/hotspot/share/gc/shared/gcLocker.cpp:109]: The `jni_lock` and `jni_unlock` naming and the matching increment/decrement gives one the impression that one is undoing the effect of the other. However, I believe that impression is incorrect. The correct value of `_jni_lock_count` is set by `set_jni_lock_count` inside a safepoint. IOW, the increment inside `jni_lock` will always be overwritten.

**Comment by albertnetymk** [comment]
> I think I agree that _debug_jni_lock_count is not really adding anything ... but perhaps there is a subtlety we are overlooking. Digging into the history of the code may shed more light.

I tracked it down to https://bugs.openjdk.java.net/browse/JDK-7129164. `_debug_jni_lock_count` tracks #threads in critical-region all the time, while `_jni_lock_count` tracks the same info but is valid only inside a safepoint. The `!needs_gc() ||` prefix seems to suggest the `_jni_lock_count` is valid only in the presence of a pending GC request. However, the currently implementation of safepoint mechanism calculates the correct value of `_jni_lock_count` unconditionally.
